optaweb-employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/solver/EmployeeRosteringConstraintProvider.java

```diff
@@ -0,0 +1,332 @@
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
+                .filter(shift -> !shift.getSpot().isCovidWard())
+                .join(Shift.class, equal(Shift::getEmployee))
+                .filter((nonCovidShift, otherShift) -> otherShift.getSpot().isCovidWard())
+                .filter((nonCovidShift, covidShift) -> covidShift.precedes(nonCovidShift))
+                .filter((nonCovidShift, covidShift) ->
+                        covidShift.getEndDateTime().until(nonCovidShift.getStartDateTime(), HOURS) < 8)
+                .ifNotExists(Shift.class,
+                        equal((nonCovidShift, covidShift) -> covidShift.getEmployee(), Shift::getEmployee),
+                        filtering((nonCovidShift, covidShift, otherShift) -> otherShift.precedes(nonCovidShift)),
+                        filtering((nonCovidShift, covidShift, otherShift) -> otherShift.follows(covidShift)))
+                .penalizeConfigurable(CONSTRAINT_COVID_NON_COVID_SHIFT_STARTED_SOON_AFTER_FINISHING_A_COVID_SHIFT);
+    }
+
+    Constraint requiredSkillForShift(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
+                .filter(shift -> !shift.getEmployee().hasSkills(shift.getSpot().getRequiredSkillSet()))
+                .penalizeConfigurable(CONSTRAINT_REQUIRED_SKILL_FOR_A_SHIFT);
+    }
+
+    Constraint unavailableEmployeeTimeSlot(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(EmployeeAvailability.class)
+                .filter(employeeAvailability -> employeeAvailability.getState() == UNAVAILABLE)
+                .join(Shift.class, equal(EmployeeAvailability::getEmployee, Shift::getEmployee))
+                .filter((employeeAvailability, shift) -> doTimeslotsIntersect(
+                        employeeAvailability.getStartDateTime(), employeeAvailability.getEndDateTime(),
+                        shift.getStartDateTime(), shift.getEndDateTime()))
+                .penalizeConfigurable(CONSTRAINT_UNAVAILABLE_TIME_SLOT_FOR_AN_EMPLOYEE);
+    }
+
+    Constraint atMostOneShiftPerDayPerEmployee(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(Shift.class,
+                equal(Shift::getEmployee),
+                equal(shift -> shift.getStartDateTime().toLocalDate()))
+                .filter((shift1, shift2) -> shift1.getEmployee() != null)
+                .penalizeConfigurable(CONSTRAINT_AT_MOST_ONE_SHIFT_ASSIGNMENT_PER_DAY_PER_EMPLOYEE);
+    }
+
+    Constraint noTwoShiftsWithin10HoursFromEachOther(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Shift.class)
+                .filter(shift -> shift.getEmployee() != null)
+                .join(Shift.class, equal(Shift::getEmployee))
```

Comment:
Add another Joiner for lessThanOrEqual to?
`Joiners.lessThanOrEqualTo(Shift::getEndTime, Shift:getStartTime)`?